packages/data-broker/src/services/Service.ts

```diff
@@ -0,0 +1,49 @@
+/**
+ * Tupaia
+ * Copyright (c) 2017 - 2020 Beyond Essential Systems Pty Ltd
+ */
+
+import {
+  AnalyticResults,
+  DataBrokerModelRegistry,
+  DataSource,
+  DataSourceType,
+  Diagnostics,
+  EventResults,
+  SyncGroupResults,
+} from '../types';
+
+export abstract class Service {
+  protected readonly models: DataBrokerModelRegistry;
+
+  public constructor(models: DataBrokerModelRegistry) {
+    this.models = models;
+  }
+
+  public get dataSourceTypes() {
+    return this.models.dataSource.getTypes();
+  }
+
+  public abstract push(
+    dataSources: DataSource[],
+    data: unknown,
+  ): Promise<{ diagnostics: Diagnostics }>;
+
+  public abstract delete(
+    dataSource: DataSource,
+    data: unknown,
+    options?: Record<string, unknown>,
+  ): Promise<Diagnostics>;
+
+  public abstract pull(
+    dataSources: DataSource[],
+    type: DataSourceType,
+    options?: Record<string, unknown>,
```

Comment:
I belive that specific schemas for `options` should not be part of this `abstract` class contract. This is different from the return type: `Promise<AnalyticResults | EventResults | SyncGroupResults | undefined> | never` since we know that a service should return either events/analytics/sync groups or throw.

To phrase it in another way, the types for this class shouldn't have to be updated every time a new service is added.

That being said, I regret the decision for using `pull` instead of `pullAnalytics` and `pulEvents`: I think it is a wrong abstraction, in the same way that forcing data elements and data groups in the same table was a wrong abstraction, I think. (`Aggregator` has a better API in that regard).